drivers/api/api.go

```diff
@@ -0,0 +1,108 @@
+package api
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/libopenstorage/openstorage/pkg/sched"
+
+	"github.com/portworx/sched-ops/task"
+	tp_errors "github.com/portworx/torpedo/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// TriggerOptions are common options used to check if any action is okay to be triggered/performed
+type TriggerOptions struct {
+	// TriggerCb is the callback function to invoke to check trigger condition
+	TriggerCb TriggerCallbackFunc
+	// TriggerCheckInterval is the interval at which to check the trigger conditions
+	TriggerCheckInterval time.Duration
+	// TriggerCheckTimeout is the duration at which the trigger checks should timeout. If the trigger
+	TriggerCheckTimeout time.Duration
+	// BlockForTrigger if set true will block the call until the TriggerCb either passes or times out
+	BlockForTrigger bool
+}
+
+// TriggerCallbackFunc is a callback function that are used by scheduler APIs to decide when to trigger/perform actions.
+// the function should return true, when it is the right time to perform the respective action
+type TriggerCallbackFunc func() (bool, error)
+
+const (
+	// defaultTriggerCheckInterval is default for the interval in trigger options
+	defaultTriggerCheckInterval = 5 * time.Second
+	// defaultTriggerCheckTimeout is the default timeout for checking the trigger options
+	defaultTriggerCheckTimeout = 5 * time.Minute
+)
+
+// PerformTask invokes the function using the given trigger options to decide when to invoke the task
+func PerformTask(userTask func() error, opts *TriggerOptions) error {
+	fn := "PerformTask"
+	if opts == nil {
+		return userTask()
+	}
+
+	f := func() error {
+		if opts.TriggerCheckTimeout == time.Duration(0) {
+			opts.TriggerCheckTimeout = defaultTriggerCheckTimeout
+		}
+
+		if opts.TriggerCheckInterval == time.Duration(0) {
+			opts.TriggerCheckInterval = defaultTriggerCheckInterval
+		}
+
+		// perform trigger checks and then perform the actual deletion
+		userTaskWithTriggerChecks := func() (interface{}, bool, error) {
+			triggered, err := opts.TriggerCb()
+			if err != nil {
+				logrus.Warnf("%s: failed to invoke trigger callback function due to: %v", fn, err)
+				return false, false, err
+			}
+
+			if triggered {
+				// run user task and return
+				logrus.Debugf("[debug] %s trigger check passed. Now running user task.\n", fn)
```

Comment:
remove this debug before commit. Use for testing.

Reply:
Done